<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.01//EN" "http://www.w3.org/TR/html4/strict.dtd"><html lang="en"><head><meta http-equiv=Content-Type content="text/html; charset=UTF-8"><style type="text/css">
body,td,div,p,a,input {font-family: arial, sans-serif;}
</style><meta http-equiv="X-UA-Compatible" content="IE=edge"><title>Gmail - Rockfish in a negative control</title><style type="text/css">
body, td {font-size:13px} a:link, a:active {color:#1155CC; text-decoration:none} a:hover {text-decoration:underline; cursor: pointer} a:visited{color:##6611CC} img{border:0px} pre { white-space: pre; white-space: -moz-pre-wrap; white-space: -o-pre-wrap; white-space: pre-wrap; word-wrap: break-word; max-width: 800px; overflow: auto;}
</style></head><body><div class="bodycontainer"><table width=100% cellpadding=0 cellspacing=0 border=0><tr height=14px><td width=143><img src="//www.gmail.com/mail/help/images/logo1.gif" width=143 height=59 alt="Gmail"></td><td align=right><font size=-1 color=#777><b>Jimmy O&#39;Donnell &lt;jodonnellbio@gmail.com&gt;</b></font></td></tr></table><hr><div class="maincontent"><table width=100% cellpadding=0 cellspacing=0 border=0><tr><td><font size=+1><b>Rockfish in a negative control</b></font><br><font size=-1 color=#777>11 messages</font></td></tr></table><hr><table width=100% cellpadding=0 cellspacing=0 border=0 class="message"><tr><td><font size=-1><b>Jesse Port </b>&lt;jport@stanford.edu&gt;</font></td><td align=right><font size=-1>Wed, Apr 22, 2015 at 11:26 AM</font><tr><td colspan=2><font size=-1 class="recipient"><div>To: Ryan Kelly &lt;rpkelly@uw.edu&gt;</div><div>Cc: Jimmy O&#39;Donnell &lt;jodonnellbio@gmail.com&gt;</div></font><tr><td colspan=2><table width=100% cellpadding=12 cellspacing=0 border=0><tr><td><div style="overflow: hidden;"><font size=-1><div dir="ltr"><div><div><div><div>Hi guys,<br><br></div>I&#39;ve been working on the paper revisions amongst other things. I&#39;m making the filtering steps as transparent as possible, providing lists of taxa excluded at each step. And an overall flowchart so the steps are clear and sequential and feed into one another. The only issue thus far is that of rockfish in one of the negative controls. This particular negative control (an extraction blank) has the least number of total sequences (~30,000), and 8,000 of them are assigned to rockfish. That seems too high of a number to enable a threshold to be set. The other two taxa in the negative controls (Human and Zoarcoidei, and I did double-check some of the Zoarcoidei sequences and they don&#39;t match human) have a total of 550,762 and 86,318 sequences respectively, so way higher. But in the actual samples, the rockfish counts vary from 0 to 67,000. So subtracting out 8,000 (if that&#39;s the approach we want to take) could make a bit of a difference. Subtracting out would also mean I&#39;ll have to revise all the figures and stats, which sounds like a major pain but if needs to be done so be it.<br><br></div>If you have any other ideas on how to handle this do let me know. Only 1 of the 3 extraction blanks had rockfish, and the collection and filtration blanks did not have rockfish either. The total counts for the negative controls (so you can see the control in question,  Ext-Blk-1F1, is at the lower end) are as follows:<br><br></div><div>Ext-Blk-1F1: 31,351<br>Ext-Blk-2F1: 461,435<br>Ext-Blk-2F3: 46,238<br></div><div>Coll-Blk: 158,851<br></div><div>Filt-Blk: 42,034<br></div><div><br></div>Thanks,<br></div>Jesse<br></div>
</font></div></table></table><hr><table width=100% cellpadding=0 cellspacing=0 border=0 class="message"><tr><td><font size=-1><b>Jimmy O&#39;Donnell </b>&lt;jodonnellbio@gmail.com&gt;</font></td><td align=right><font size=-1>Wed, Apr 22, 2015 at 5:58 PM</font><tr><td colspan=2><font size=-1 class="recipient"><div>To: Jesse Port &lt;jport@stanford.edu&gt;</div><div>Cc: Ryan Kelly &lt;rpkelly@uw.edu&gt;</div></font><tr><td colspan=2><table width=100% cellpadding=12 cellspacing=0 border=0><tr><td><div style="overflow: hidden;"><font size=-1><div dir="ltr">Hey Jesse,<div><br></div><div>Ryan&#39;s in Monaco (gambling? polishing his yacht? not exactly sure...), so he might be slow to respond.</div><div><br></div><div>Just to make sure I follow correctly, and to put it in terms that are taxon-independent, are these the options we&#39;re considering?</div><div><br></div><div>OPTION 1:</div><div>IF: a sequence is found is the negative controls</div><div>THEN: remove all occurrences of that sequence from all other samples.</div><div><br></div><div>OPTION 2:</div><div>IF: a sequence is found in the negative controls</div><div>THEN: calculate its (relative? | raw?) abundance, and subtract up to that amount from the respective value in other samples.</div><div><br></div><div>Option 1 seems way too conservative, I just threw it out to be clear. I&#39;d guess we&#39;re all on board for one of two flavors of option 2 (remove the corresponding amount of relative or raw number of reads from other samples). I don&#39;t think it should affect the results that much. For example, in the multi-panel figure, it would bring down the total number of reads within a taxa; but, that won&#39;t make a difference as the comparisons are are only within a taxon among samples.</div><div><br></div><div>Regarding the pain in the ass work of redoing figures and stats -- I&#39;m not trying to rub salt in any wounds, but if we write this in code (I&#39;m actually working on this right now), it should be absolutely trivial.</div><div><br></div><div>Let me know how I can help!</div><div><br></div><div>Jimmy</div><div><br></div></div><div class="gmail_extra"><br clear="all"><div><div><div dir="ltr"><div><div dir="ltr"><div><span style="font-size:small">______________________________<wbr>_____</span><font color=#888888><br></font></div><font color=#888888><div>Jimmy O&#39;Donnell<br>Research Associate</div><div>School of Marine and Environmental Affairs</div><div>University of Washington</div><div>Seattle, WA 98105</div><div><a href="mailto:jodonnellbio@gmail.com" target="_blank">jodonnellbio@gmail.com</a><br><a href="tel:%28407%29744-3377" value="+14077443377" target="_blank">(407)744-3377</a></div></font></div></div></div></div></div><div style='padding:5 0'><font size=1 color=#888888>[Quoted text hidden]</font></div></div>
</font></div></table></table><hr><table width=100% cellpadding=0 cellspacing=0 border=0 class="message"><tr><td><font size=-1><b>Ryan Kelly </b>&lt;rpkelly@uw.edu&gt;</font></td><td align=right><font size=-1>Wed, Apr 22, 2015 at 7:48 PM</font><tr><td colspan=2><font size=-1 class="recipient"><div>To: Jimmy O&#39;Donnell &lt;jodonnellbio@gmail.com&gt;</div><div>Cc: Jesse Port &lt;jport@stanford.edu&gt;</div></font><tr><td colspan=2><table width=100% cellpadding=12 cellspacing=0 border=0><tr><td><div style="overflow: hidden;"><font size=-1><div style="word-wrap:break-word">Hi from Monaco, where I can’t sleep for some reason. <div><br></div><div>Is it worth reviewing what the Taberlet group would do here, so we can cite their method?</div><font color=#888888><div> </div></font><div><font color=#888888>-r</font><div style='padding:5 0'><font size=1 color=#888888>[Quoted text hidden]</font></div></div></div></font></div></table></table><hr><table width=100% cellpadding=0 cellspacing=0 border=0 class="message"><tr><td><font size=-1><b>Jimmy O&#39;Donnell </b>&lt;jodonnellbio@gmail.com&gt;</font></td><td align=right><font size=-1>Wed, Apr 22, 2015 at 9:56 PM</font><tr><td colspan=2><font size=-1 class="recipient"><div>To: Ryan Kelly &lt;rpkelly@uw.edu&gt;</div><div>Cc: Jesse Port &lt;jport@stanford.edu&gt;</div></font><tr><td colspan=2><table width=100% cellpadding=12 cellspacing=0 border=0><tr><td><div style="overflow: hidden;"><font size=-1><div dir="ltr">I&#39;m all for that.</div><div class="gmail_extra"><font color=#550055><br clear="all"><div><div><div dir="ltr"><div><div dir="ltr"><div><span style="font-size:small">______________________________<wbr>_____</span><br></div><div>Jimmy O&#39;Donnell<br>Research Associate</div><div>School of Marine and Environmental Affairs</div><div>University of Washington</div><div>Seattle, WA 98105</div><div><a href="mailto:jodonnellbio@gmail.com" target="_blank">jodonnellbio@gmail.com</a><br><a href="tel:%28407%29744-3377" value="+14077443377" target="_blank">(407)744-3377</a></div></div></div></div></div></div>
<br></font><div style='padding:5 0'><font size=1 color=#888888>[Quoted text hidden]</font></div></div>
</font></div></table></table><hr><table width=100% cellpadding=0 cellspacing=0 border=0 class="message"><tr><td><font size=-1><b>Jesse Port </b>&lt;jport@stanford.edu&gt;</font></td><td align=right><font size=-1>Thu, Apr 23, 2015 at 12:05 PM</font><tr><td colspan=2><font size=-1 class="recipient"><div>To: Jimmy O&#39;Donnell &lt;jodonnellbio@gmail.com&gt;</div><div>Cc: Ryan Kelly &lt;rpkelly@uw.edu&gt;</div></font><tr><td colspan=2><table width=100% cellpadding=12 cellspacing=0 border=0><tr><td><div style="overflow: hidden;"><font size=-1><div dir="ltr"><div><div><div><div><div><div><div>Hey guys,<br><br></div>Thanks for getting back. Forgot you were in Monaco Ryan. Are you in fact polishing your yacht?<br><br></div>With regard to the Taberlet approach for filtering of negative controls, here&#39;s what they did in the De Barba paper (PCR- = negative controls):<br><br>&quot;Examination of consensus sequence profiles obtained for any given sample by averaging read counts of the remaining sequences among replicates showed that<br>PCR- had a total per sample counts considerably smaller than other samples (Fig. 3g-h). We took this as an indication that faecal samples with small total read<br>counts could reasonably be considered of low DNA quantity for the marker under analysis and therefore more easily exposed to random amplification of the few target molecules present in the template. Following this reasoning, we deleted samples with total mean consensus read count &lt;1000 for all universal and &lt;500 for all ITS markers, which resulted in the removal of all PCR-, and a variable number of faecal samples in all markers data (Table 2).&quot;<br><br></div>So all their negative controls were discarded using this filtering. If we took the same approach, we also have a big discrepancy between the total number of reads in the negative controls (mean=26,700 excluding human sequences) vs. field samples (mean=195,000), but much higher counts in the negative controls. I excluded human sequences because they just dominate the controls and inflate the count total.<br><br></div>So the question is, could we still use the approach of discarding the negative controls based on some threshold of total counts in the negative controls in relation to total counts for field samples? Or is the raw count value already too high to use this approach? If so, I think we&#39;re stuck with having to use the subtraction method and subtract out the raw count value of a taxon in the negative control from the field samples. Which unfortunately means redoing figures/stats just to account for the change in rockfish counts. Shouldn&#39;t change much but need to make sure the data matches the figures/numbers. And yes, we should definitely get this all coded.<br><br></div>Let me know what you think.<br><br></div>Thanks,<br></div>Jesse<br></div><div class="gmail_extra"><div style='padding:5 0'><font size=1 color=#888888>[Quoted text hidden]</font></div><font color=#888888>-- <br><div><div dir="ltr"><div><div dir="ltr"><div><div dir="ltr">Jesse Port, PhD<br>Stanford University<br>Early Career Fellow, Center for Ocean Solutions<br><a href="mailto:jport@stanford.edu" target="_blank">jport@stanford.edu</a><br><font size="2"><span style="text-indent:0px;letter-spacing:normal;font-variant:normal;font-style:normal;font-weight:normal;line-height:normal;border-collapse:separate;text-transform:none;white-space:normal;font-family:Helvetica;word-spacing:0px"><span style="text-indent:0px;letter-spacing:normal;font-variant:normal;font-style:normal;font-weight:normal;line-height:normal;border-collapse:separate;text-transform:none;white-space:normal;font-family:Helvetica;word-spacing:0px"></span></span></font><br></div></div></div></div></div></div>
</font></div>
</font></div></table></table><hr><table width=100% cellpadding=0 cellspacing=0 border=0 class="message"><tr><td><font size=-1><b>Jimmy O&#39;Donnell </b>&lt;jodonnellbio@gmail.com&gt;</font></td><td align=right><font size=-1>Thu, Apr 23, 2015 at 8:46 PM</font><tr><td colspan=2><font size=-1 class="recipient"><div>To: Jesse Port &lt;jport@stanford.edu&gt;</div><div>Cc: Ryan Kelly &lt;rpkelly@uw.edu&gt;</div></font><tr><td colspan=2><table width=100% cellpadding=12 cellspacing=0 border=0><tr><td><div style="overflow: hidden;"><font size=-1><div dir="ltr">Yo Jesse,<div><br></div><div>I take from this that they didn&#39;t do any taxon-based filtering. Seems fair, and works in our favor. Their thresholds (1000 and 500) seem completely arbitrary; I don&#39;t see why we shouldn&#39;t follow suit. I would not make any distinction based on taxa in setting this threshold -- that is, leave the human sequences in, report the total number of reads per sample, and set the threshold for something greater than the maximum reads per negative control.</div><div><br></div><div>Judging from this figure (attached), that threshold should be somewhere around 1.2e5, right? I think it&#39;s ok to say &quot;one of our extraction blanks was clearly screwed up -- the majority of those reads were human (X%), so we assumed there was human contamination in the lab in that sample. We fired that tech and excluded that sample from further analysis.&quot;<br></div><div><br></div><div>Unless you want to do taxon-based filtering (i.e. trim a certain number of reads from field-derived samples dependent on how many of those sequences were found in the controls), I don&#39;t think we have to. (Unless the reviewer specifically asked for this?)<br></div><div><br></div><div>Just to clarify, did you mean POSITIVE here: &quot;...<span style="font-size:12.8000001907349px"> </span><span style="font-size:12.8000001907349px">vs. field samples (mean=195,000), but much higher counts in the negative controls.</span><span style="font-size:12.8000001907349px"> &quot;</span></div><div><br></div><div>Here&#39;s a photo Ryan sent of him standing in front of his shiny yacht in Monaco. Looks great Ryan!</div><div><br></div><div><img src="https://ci6.googleusercontent.com/proxy/--yfzFha5CJffXRPYyuGBfWYjdvjKl_7NZCEIlPq6vendBwmXgixOYxHge1hKYFJpK60CSBwVF4v6sXOe5vEQ-ro7nXPqUwja0WJ9Pw4LQo=s0-d-e1-ft#http://www.wolfgangbrinck.com/boats/restoration/bhoney.jpg" width="476" height="287"><font color=#888888><br><br></font></div><font color=#888888><div><br></div><div>Jimmy</div><div><br></div></font></div><div class="gmail_extra"><font color=#550055><br clear="all"><div><div><div dir="ltr"><div><div dir="ltr"><div><span style="font-size:small">______________________________<wbr>_____</span><br></div><div>Jimmy O&#39;Donnell<br>Research Associate</div><div>School of Marine and Environmental Affairs</div><div>University of Washington</div><div>Seattle, WA 98105</div><div><a href="mailto:jodonnellbio@gmail.com" target="_blank">jodonnellbio@gmail.com</a><br><a href="tel:%28407%29744-3377" value="+14077443377" target="_blank">(407)744-3377</a></div></div></div></div></div></div>
<br></font><div style='padding:5 0'><font size=1 color=#888888>[Quoted text hidden]</font></div></div>
</font></div><br clear=all><div style="width:50%;border-top:2px #AAAAAA solid"></div><table class=att cellspacing=0 cellpadding=5 border=0><tr><td><table cellspacing=0 cellpadding=0><tr><td><a target=_blank href="?ui=2&amp;ik=4174f4c421&amp;view=att&amp;th=14ce989076a6492e&amp;attid=0.1&amp;disp=attd&amp;realattid=f_i8v1xvd10&amp;safe=1&amp;zw"><img width=16 height=16 src="https://ssl.gstatic.com/ui/v1/icons/mail/images/pdf.gif"></a></td><td width=7></td><td><b>reads_per_sample_5_groups.pdf</b><br/>5K </td></tr></table></td></tr></table></table></table><hr><table width=100% cellpadding=0 cellspacing=0 border=0 class="message"><tr><td><font size=-1><b>Jesse Port </b>&lt;jport@stanford.edu&gt;</font></td><td align=right><font size=-1>Fri, Apr 24, 2015 at 10:29 AM</font><tr><td colspan=2><font size=-1 class="recipient"><div>To: Jimmy O&#39;Donnell &lt;jodonnellbio@gmail.com&gt;</div><div>Cc: Ryan Kelly &lt;rpkelly@uw.edu&gt;</div></font><tr><td colspan=2><table width=100% cellpadding=12 cellspacing=0 border=0><tr><td><div style="overflow: hidden;"><font size=-1><div dir="ltr"><div>Oh nice Ryan, I like the green/brown coloration! You really are a jack of all trades.<br><br></div><div>Thanks for your input Jimmy. I&#39;d like to go with the threshold approach (based on the max reads in any of the negative controls) for discarding the negative controls, but it seems a bit awkward given the fact that we know the negative controls do contain obvious contamination that also appears in the field samples (human, Zoarcoidei, and to a much lesser extent rockfish). So if we toss out the negative controls because their total read counts were below a threshold, we have no way of actually discarding these taxa from the field samples even though we know that the negative controls confirm that they are contamination. So that makes me think taxon-based filtering is the only way to remove the contaminant taxa from the field samples. <br><br>This make sense or do you still think we can rationalize a threshold approach?<br></div></div><div style='padding:5 0'><font size=1 color=#888888>[Quoted text hidden]</font></div></font></div></table></table><hr><table width=100% cellpadding=0 cellspacing=0 border=0 class="message"><tr><td><font size=-1><b>Ryan Kelly </b>&lt;rpkelly@uw.edu&gt;</font></td><td align=right><font size=-1>Fri, Apr 24, 2015 at 10:48 AM</font><tr><td colspan=2><font size=-1 class="recipient"><div>To: Jesse Port &lt;jport@stanford.edu&gt;</div><div>Cc: Jimmy O&#39;Donnell &lt;jodonnellbio@gmail.com&gt;</div></font><tr><td colspan=2><table width=100% cellpadding=12 cellspacing=0 border=0><tr><td><div style="overflow: hidden;"><font size=-1><div style="word-wrap:break-word"><div>Could we do both?  The logic would be as follows:</div><div><br></div><div>1. Samples with very low numbers of reads are likely to be unreliable/spurious.</div><div>2. Species that dominate the controls but are more rare in the field samples are likely to be due to contamination. This is particularly the case for non-marine species.</div><div>3. As a result, we first removed reads from all samples for taxa that occurred primarily in the negative controls (i.e., where more than some large percentage—e.g. 90%—of reads for a given taxon, considering the dataset as a whole, were in the controls).</div><div>4. We then removed samples with low total numbers of reads.</div><div><br></div><div>Would this solve the problem, or does it still leave us with a bunch of human in the samples? And even if it does, that’s probably OK, no?</div><font color=#888888><div><br></div><div> -r</div></font><div style='padding:5 0'><font size=1 color=#888888>[Quoted text hidden]</font></div></div></font></div></table></table><hr><table width=100% cellpadding=0 cellspacing=0 border=0 class="message"><tr><td><font size=-1><b>Ryan Kelly </b>&lt;rpkelly@uw.edu&gt;</font></td><td align=right><font size=-1>Fri, Apr 24, 2015 at 10:49 AM</font><tr><td colspan=2><font size=-1 class="recipient"><div>To: Jesse Port &lt;jport@stanford.edu&gt;</div><div>Cc: Jimmy O&#39;Donnell &lt;jodonnellbio@gmail.com&gt;</div></font><tr><td colspan=2><table width=100% cellpadding=12 cellspacing=0 border=0><tr><td><div style="overflow: hidden;"><font size=-1><div style="word-wrap:break-word"><div>also, Monaco is ridiculous.</div><div><br></div><img height="240" width="320" src="?ui=2&amp;ik=4174f4c421&amp;view=fimg&amp;th=14cec8c9da69f10b&amp;attid=0.1.1&amp;disp=emb&amp;attbid=ANGjdJ-8LuwYlSkvU4gMzBML78ekUrEhhZfmxLPS59Gys0O1ViQ8nooFvP2SqsP0pW8OpF19_-_0xAaptx_5oFOpHGUpXkt8T8BkkjlN92X4JqBia-fe26oEoEXx5kw&amp;sz=w640-h480&amp;ats=1436321881609&amp;rm=14cec8c9da69f10b&amp;zw&amp;atsh=1"><img height="240" width="320" src="?ui=2&amp;ik=4174f4c421&amp;view=fimg&amp;th=14cec8c9da69f10b&amp;attid=0.1.2&amp;disp=emb&amp;attbid=ANGjdJ8JQpGfK0hAlkhphMiwuimRgcI37Y7HzfKScCgC0PnS-0gOetXCjfOFqTPSEwGog20oXhrAA4M3_j0FdlFGAs6c9xR11Z3YThyQaWIjQEOzNXRZTUZ5rHkyKeU&amp;sz=w640-h480&amp;ats=1436321881609&amp;rm=14cec8c9da69f10b&amp;zw&amp;atsh=1"><div><img height="240" width="320" src="?ui=2&amp;ik=4174f4c421&amp;view=fimg&amp;th=14cec8c9da69f10b&amp;attid=0.1.3&amp;disp=emb&amp;attbid=ANGjdJ_xGUy9-OIxGfZzj9_PJ-amiLKs6KS9dPO7b_gOU6elNhyPZK8nxJiZ3V4M494GT09IZ-s1l1Ua0Put4frYhT2qOstgfR6ZlRlXbk0UQZFb1E8BC_3Xa27xpTg&amp;sz=w640-h480&amp;ats=1436321881610&amp;rm=14cec8c9da69f10b&amp;zw&amp;atsh=1"><img height="240" width="320" src="?ui=2&amp;ik=4174f4c421&amp;view=fimg&amp;th=14cec8c9da69f10b&amp;attid=0.1.4&amp;disp=emb&amp;attbid=ANGjdJ9JyJhRrTQhUBCybPN-9UZgph58fBIwNPvmaX0WQzKsLh11ekNPz4_niIYeXsBGKMW5HDyG__gNXQETonUCJmpdynhPDb0YG7c04hZwfJs2gpTg8_jMS3mQJRU&amp;sz=w640-h480&amp;ats=1436321881610&amp;rm=14cec8c9da69f10b&amp;zw&amp;atsh=1"><div style='padding:5 0'><font size=1 color=#888888>[Quoted text hidden]</font></div></div></div></font></div></table></table><hr><table width=100% cellpadding=0 cellspacing=0 border=0 class="message"><tr><td><font size=-1><b>Jesse Port </b>&lt;jport@stanford.edu&gt;</font></td><td align=right><font size=-1>Fri, Apr 24, 2015 at 11:46 AM</font><tr><td colspan=2><font size=-1 class="recipient"><div>To: Ryan Kelly &lt;rpkelly@uw.edu&gt;</div><div>Cc: Jimmy O&#39;Donnell &lt;jodonnellbio@gmail.com&gt;</div></font><tr><td colspan=2><table width=100% cellpadding=12 cellspacing=0 border=0><tr><td><div style="overflow: hidden;"><font size=-1><div dir="ltr"><div><span style="color:rgb(0,0,255)"><span style="color:rgb(0,0,0)">Nice pics Ryan ;) Regarding your comments:</span><font color=#550055><br><br>1. Samples with very low numbers of reads are likely to be unreliable/spurious.</font></span><br></div><div>None of the controls have very low numbers of reads (min=31,000) so not really applicable. Taberlet&#39;s minimum was 1000-5000 reads.<br><br></div><font color=#550055><div><span style="color:rgb(0,0,255)">2.
Species that dominate the controls but are more rare in the field
samples are likely to be due to contamination. This is particularly the
case for non-marine species.</span><br></div></font><div>This is the case for Zoarcoidei but not human or rockfish.<br></div><font color=#550055><div><br></div><div><span style="color:rgb(0,0,255)">3. As a result, we first removed
reads from all samples for taxa that occurred primarily in the negative
controls (i.e., where more than some large percentage—e.g. 90%—of reads
for a given taxon, considering the dataset as a whole, were in the
controls).</span><br></div></font><div>Per above, this would remove Zoarcoidei but we&#39;d still be left with human and rockfish. And I don&#39;t think this could be at the level of the whole dataset considering there are 36 field samples and 5 negative controls. But one of the controls has a much higher number of human reads (360K) than the max of any field sample (147K). We could consider comparing the average of the taxa across the field samples vs. controls. For human this would be 46K field vs. 110K control. For Zoarcoidei this would be 4K field vs. 43K control. For rockfish 20K field vs. 1600K controls, so doesn&#39;t work for rockfish. But rockfish could potentially become a non-issue if there is a reasonable way to do the next step.<br><br></div><font color=#550055><div><span style="color:rgb(0,0,255)">4. We then removed samples with low total numbers of reads.</span><br></div></font><div>This gets back to how we define low, and even if we set a cutoff at 100K reads, we&#39;d retain human in the dataset which we know is obvious contamination because it was abundant in the negative control. Unless we consider the mean comparisons I just described, then human is gone before this step.<br><br></div><div>If this all sounds too complicated or convoluted I can just do taxon-based filtering. It&#39;ll take a bit to get the figures and stats redone.<font color=#888888><br><br></font></div><font color=#888888><div>jesse<br></div></font></div><div style='padding:5 0'><font size=1 color=#888888>[Quoted text hidden]</font></div></font></div></table></table><hr><table width=100% cellpadding=0 cellspacing=0 border=0 class="message"><tr><td><font size=-1><b>Ryan Kelly </b>&lt;rpkelly@uw.edu&gt;</font></td><td align=right><font size=-1>Fri, Apr 24, 2015 at 2:09 PM</font><tr><td colspan=2><font size=-1 class="recipient"><div>To: Jesse Port &lt;jport@stanford.edu&gt;</div><div>Cc: Jimmy O&#39;Donnell &lt;jodonnellbio@gmail.com&gt;</div></font><tr><td colspan=2><table width=100% cellpadding=12 cellspacing=0 border=0><tr><td><div style="overflow: hidden;"><font size=-1><div style="word-wrap:break-word"><div>Yeah - tricky. One thought I had was to think through why we have controls in the first place: what are we controlling for?</div><div><br></div><div>I think we do/did negative controls to ensure we can distinguish sequences in the field from contamination (this is seemingly straightforward, but stay with me). This leaves us with the issue that any species that shows up in the negative controls AND in the field samples might be an artifact of contamination. So, human, rockfish, and a bunch of species that all have a common ancestor within Zoarcoidei (eel-like fish of which several are in Monterey).</div><div><br></div><div>By contrast, Taberlet et al. seem to be using the negative controls as controls on the sequencing reaction—essentially, to determine if the sample runs yield reads at a rate greater than a blank. This allows them to identify and remove low-quality samples as likely to be subject to contamination. But that doesn’t really apply in our case, now that I think about it, because we don’t have low-read samples. </div><div><br></div><div>We could use the controls in another way — and we’ve seen this somewhere before — which is as a yardstick for sequencing/annotation at the community level. So, we do a PERMANOVA (adonis), and anything that’s not significantly different from the controls, we throw out.  Anything that is significantly different from the controls, we keep.  That solves much of the problem, but leaves us with some taxa (mostly human, but some others) we think are spurious in the samples. We could then do a secondary sweep, removing taxa that a) are above a critical threshold of abundance in the controls, and/or b) some other reasonable criterion. But does this remove rockfish?</div><font color=#888888><div><br></div><div> -r</div></font><div style='padding:5 0'><font size=1 color=#888888>[Quoted text hidden]</font></div></div></font></div></table></table></div></div><script type="text/javascript">// <![CDATA[
// document.body.onload=function(){document.body.offsetHeight;window.print()};
// ]]></script></body></html>